### ISSUE TITLE
DataContractSerializer failed with some Monogame types

### ISSUE THREAD

**Issue by Bletraut** [MonoGame/MonoGame#8508]
### Prerequisites

- [X] I have verified this issue is present in the `develop` branch
- [X] I have searched [open](https://github.com/MonoGame/MonoGame/issues) and [closed](https://github.com/MonoGame/MonoGame/issues?q=is%3Aissue+is%3Aclosed) issues to ensure it has not already been reported.

### MonoGame Version

MonoGame v3.8.2

### Which MonoGame platform are you using?

MonoGame Cross-Platform Desktop Application (mgdesktopgl)

### Operating System

Windows

### Description

I found several types that serialize with errors when using **DataContractSerializer**.
For example, you can't serialize a property that doesn't have a getter. You can't serialize indexers either. Exceptions can't also implement ISerializable.

I suggest removing the **DataContract** attribute for exceptions, and also placing the **DataMember** attribute on private fields instead of properties without getters, which cause an error with **DataContractSerializer**.

```csharp
    [DataContract]
    public class Curve : ICurveEvaluator<float>
    {
        #region Private Fields

        private CurveLoopType _preLoop;
        private CurveLoopType _postLoop;
        // Place DataMember here and serialize private field.
        [DataMember]
        private CurveKeyCollection _keys;

        #endregion

        #region Public Properties

        /// <summary>
        /// Returns <c>true</c> if this curve is constant (has zero or one points); <c>false</c> otherwise.
        /// </summary>
        // Remove DataMember from here because it throws an exception.
        [DataMember]
        public bool IsConstant
        {
            get { return this._keys.Count <= 1; }
        }
        ...
     }
```

If this is a good suggestion, I will create a PR with corrections.

### Steps to Reproduce

Create a empty project and try serialize a Scene with **DataContractSerializer**.

```csharp
public class Scene {}

public void SaveScene(string path)
{
    var knownTypes = AppDomain.CurrentDomain.GetAssemblies()
        .SelectMany(assembly => assembly.GetTypes().Where(type => type.IsDefined(typeof(DataContractAttribute))));

    var serializer = new DataContractSerializer(typeof(Scene), knownTypes);
    using var writer = XmlWriter.Create(path);
    serializer.WriteObject(writer, _scene);
}
```

### Expected Behavior

All data is serialized.

### Resulting Behavior

Some types throw errors when serializing.
I found these broken types. Please note that this does not include all exceptions for a type.
**Curve** throws
```
System.Runtime.Serialization.InvalidDataContractException: 
'No set method for property 'IsConstant' in type 'Microsoft.Xna.Framework.Curve'.'
```

**CurveKey** throws
```
System.Runtime.Serialization.InvalidDataContractException: 
'No set method for property 'Position' in type 'Microsoft.Xna.Framework.CurveKey'.'
```

**CurveKeyCollection** throws
```
System.Runtime.Serialization.InvalidDataContractException: 
'Property 'Item' in type 'Microsoft.Xna.Framework.CurveKeyCollection' cannot be serialized
 because serialization of indexed properties is not supported.'
```

**DeviceLostException, DeviceNotResetException, NoSuitableGraphicsDeviceException, InstancePlayLimitException, NoAudioHardwareException, NoMicrophoneConnectedException** throw
```
System.Runtime.Serialization.InvalidDataContractException: 
'Type 'Microsoft.Xna.Framework.Graphics.DeviceLostException' cannot be ISerializable and have
 DataContractAttribute attribute.'
```

**Comment by stromkos** [comment]
Why???

Games only serialize there own data state for save/restore of games.

> System.Runtime.Serialization.InvalidDataContractException: 
'Type 'Microsoft.Xna.Framework.Graphics.DeviceLostException' cannot be ISerializable and have
 DataContractAttribute attribute.'

Nothing in "Microsoft.Xna.Framework.Graphics" should be saved.

**Comment by Bletraut** [comment]
> Why???
> 
> Games only serialize there own data state for save/restore of games.
> 
> > System.Runtime.Serialization.InvalidDataContractException:
> > 'Type 'Microsoft.Xna.Framework.Graphics.DeviceLostException' cannot be ISerializable and have
> > DataContractAttribute attribute.'
> 
> Nothing in "Microsoft.Xna.Framework.Graphics" should be saved.

Yes, you are right. It is unlikely that anyone will use these classes for serialization. But they also have an erroneously set DataContract attribute.

**Comment by dellis1972** [comment]
Closing this issue as these types should not be used in game save data. You can use the `IgnoreDataMember` attribute if you want to exclude certain fields/properties from your own serialisation types.
